### DISCUSSION TITLE
Dagger Universe current state and proposal for new package submission and promotion

### DISCUSSION THREAD

**Discussion by jpadams** [dagger/dagger#2616] [General]
### Hello Dagger Community! 👋
cc @dagger/maintainers 

We've been rethinking contribution of content to Dagger Universe and want to share this proposal for comment as we refine the plan. I'll start with the current state and our proposed future.

### Up until recently:
There were two kinds of Dagger Universe packages:
- official ones under `universe.dagger.io/<package name>`
- experimental ones under `universe.dagger.io/x/<email address>/<package name>`

Well, maybe three...In addition there were packages that folks kept in their own git repos and added to their Dagger projects via `dagger project update https://<git server>/<git username>/<package name>@<ref>`

### Very, very recently:
We've created `universe.dagger.io/alpha` and `universe.dagger.io/beta` to be the promotion path for new Dagger Universe packages on their way to being an official stable package.

```mermaid
graph LR
    A(alpha) --> B(beta) --> C(stable)
```

We've decided to do away with `universe.dagger.io/x/<email address>/*` and have made a one-time move of those packages to `universe.dagger.io/alpha/*` with some consolidation of related packages (under `kubernetes` for example). So, we won't take any more PRs for those packages, but will leave them in place for backwards compatibility for a period of time. We'll also mark them as `//Deprecated: in favor of universe.dagger.io/alpha/<package name>` in comments in the `.cue` files.

**Going forward, please make new package submissions via PR to `universe.dagger.io/alpha/<package name>`.**

### Proposed Next:
New packages will start in `alpha`. Once a package has been merged into `alpha`, the author would be added to the `CODEOWNERS` file for `universe.dagger.io/alpha/<package name>/` (and same for `beta` later) to help maintain the package.

When a package has matured, a PR can be opened to promote it from `alpha` to `beta`. One sign of maturity is adequate test coverage of the package. There should be a minimum of 50% test coverage at this stage, though 80%+ is preferred. `//Deprecated: in favor of universe.dagger.io/beta/<package name>` will be added.

For a package to move from `beta` -> stable (just plain `universe.dagger.io/<package name>`, there should be additional scrutiny around quality (style, comments, conventions, security), the package should have 80%+ test coverage, and usage metrics could help indicate usefulness. Some curation/discussion will likely occur at this stage to balance the desire for diversity/quantity of packages with the need for high-quality packages that deliver a great developer experience. `//Deprecated: in favor of universe.dagger.io/<package name>` will be added.

We expect that in the near future there will be a Dagger Universe catalog of packages, so metadata like a description, example usage, etc will be needed in a separate PR as well (TBD).

Automation will be important to prevent making updates to deprecated packages and to test for quality and test coverage. Automation could open PRs for promotion from `alpha`->`beta` when a quality bar is met and could add `//Deprecated: xxx` comments, etc.

#### Thank you! @jpadams

**Comment by TomChv** [comment]
That's a great proposal! I have few questions:

1 - What about current PRs (#2452, #2533, #2581 and #2578) ? Should they update it or we take care about copying those to `alpha`?

2 - I do not see `kubectl` package in Alpha, what is the decision for packages that has multiple implementation?

3 - How will we test packages that interact with third-parties (`aws/sam`, `discord`) or required infrastructure (`kubernetes`, `pulumi`) ? Is it up to dagger to set infrastructure or to the user? This will have a big impact on CI time

4 - Is there any tool to see the coverage of tests in CUE? How we can say "this package has 50% coverage, this one 80%",

5 - Who is responsible of the maintenance of a package when it's populate to official Dagger universe?

6 - Totally agree, do you have any idea how we can take care of this? 

> Automation will be important to prevent making updates to deprecated packages and to test for quality and test coverage. Automation could open PRs for promotion from alpha->beta when a quality bar is met and could add //Deprecated: xxx comments, etc.

**Reply by jpadams** [reply]
> That's a great proposal! I have few questions:
> 
> 1 - What about current PRs (#2452, #2533, #2581 and #2578) ? Should they update it or we take care about copying those to `alpha`?

We should redirect those PRs to `alpha`

> 2 - I do not see `kubectl` package in Alpha, what is the decision for packages that has multiple implementation?

https://github.com/dagger/dagger/pull/2579#issue-1260905491
Packages with competing implementations were not copied yet. We'll need to work with the authors (including you) and the team to decide on an implementation for `alpha`.

> 3 - How will we test packages that interact with third-parties (`aws/sam`, `discord`) or required infrastructure (`kubernetes`, `pulumi`) ? Is it up to dagger to set infrastructure or to the user? This will have a big impact on CI time

I'm guessing we will have to make exceptions in some cases, or use API mocking, or find some other creative solution.

> 4 - Is there any tool to see the coverage of tests in CUE? How we can say "this package has 50% coverage, this one 80%",

We don't have a tool selected yet. This will need to be worked out, though many tools exist for this purpose.

> 5 - Who is responsible of the maintenance of a package when it's populate to official Dagger universe?

Since it will continue to be open source, the whole community will continue to support the package.

> 6 - Totally agree, do you have any idea how we can take care of this?
> > Automation will be important to prevent making updates to deprecated packages and to test for quality and test coverage. Automation could open PRs for promotion from alpha->beta when a quality bar is met and could add //Deprecated: xxx comments, etc.

We'll put our heads together and likely adapt something.

**Comment by sagikazarmark** [comment]
Personally, I'd keep a final gate in the form of manual approve from a maintainer before promoting something to prod.

There is also the question of versioning (or backward compatibility promise). Since everything is in a single, giant repo it seems quite hard to do versioning. So anything that gets promoted to prod (or even beta) needs to stay backward compatible....indefinitely? At the very least we need to specify the requirements around backward compatibility.

After writing a couple actions, I don't feel very confident (yet) about certain patterns and practices I applied, so I'd definitely wait some time before applying the promotion process. Just until enough patterns emerge from building and using actions.
